### PR TITLE
add http client integration tests

### PR DESCRIPTION
**Pull request by hoanhan101**
related to #63

### CODE REVIEW

File: synse/http_integration_test.go [hunk 1]
```diff
@@ -0,0 +1,551 @@
...
+	assert.Equal(t, 1, len(plugins))
+
+	for _, plugin := range plugins {
+		assert.NotEmpty(t, plugin.Name)
+		assert.NotEmpty(t, plugin.Maintainer)
+		assert.NotEmpty(t, plugin.Tag)
+		assert.NotEmpty(t, plugin.Description)
+		assert.NotEmpty(t, plugin.ID)
+		assert.True(t, plugin.Active)
+	}
```
Comment (edaniszewski) [inline synse/http_integration_test.go]: I think that since we control the integration test environment and know that there should only ever be the one emulator plugin, instead of iterating through the slice and doing a generic check, we could do checks specific to the emulator, e.g.

```go
p := plugins[0]
assert.Equal(t, "emulator", plugin.Name)
assert.Equal(t, "vaporio", plugin.Maintainer)
...
```

File: synse/http_integration_test.go [hunk 2]
```diff
@@ -0,0 +1,551 @@
...
+		assert.NotEmpty(t, plugin.Version.SDKVersion)
+		assert.NotEmpty(t, plugin.Version.BuildDate)
+		assert.NotEmpty(t, plugin.Version.GitCommit)
+		assert.NotEmpty(t, plugin.Version.GitTag)
+		assert.NotEmpty(t, plugin.Version.Arch)
+		assert.NotEmpty(t, plugin.Version.OS)
+		assert.NotEmpty(t, plugin.Health.Timestamp)
+		assert.Equal(t, "OK", plugin.Health.Status)
+		assert.Equal(t, 2, len(plugin.Health.Checks))
+
```
Comment (edaniszewski) [inline synse/http_integration_test.go]: same note as above, we can make some of these checks more specific to the plugin that we know we are testing against.

I think the `NotEmpty` check is still good for things like version info/build date, etc since they are liable to change, but other info is more static (name, maintainer, tag, ...)

File: synse/http_integration_test.go [hunk 3]
```diff
@@ -0,0 +1,551 @@
...
+	client, err := NewHTTPClientV3(&Options{
+		Address: "localhost:5000",
+	})
+	assert.NotNil(t, client)
+	assert.NoError(t, err)
+
+	health, err := client.PluginHealth()
+	assert.NoError(t, err)
+	assert.NotEmpty(t, health.Status)
```
Comment (edaniszewski) [inline synse/http_integration_test.go]: does this need to be a NotEmpty check? If we can check for `len(health.Healthy)` and `health.Active`, etc, then I imagine the health status should also be populated with a known value?

File: synse/http_integration_test.go [hunk 4]
```diff
@@ -0,0 +1,551 @@
...
+	assert.Equal(t, 22, len(devices))
+
+	for _, device := range devices {
+		assert.NotEmpty(t, device.ID)
+		assert.NotEmpty(t, device.Info)
+		assert.NotEmpty(t, device.Type)
+		assert.NotEmpty(t, device.Plugin)
+		assert.NotEmpty(t, device.Tags)
+
+		// only led devices have alias
```
Comment (edaniszewski) [inline synse/http_integration_test.go]: I'm not sure I'd hardcode this check into the test, as this is liable to change given a simple change to the plugin configuration. If we run the integration tests using a custom device configuration where we can guarantee that this wont change, then I think this would be fine, but since it appears we are just using the default baked-in config, I think this may lead to issues down the road when it inevitability changes.

File: synse/http_integration_test.go [hunk 5]
```diff
@@ -0,0 +1,551 @@
...
+	}
+
+	client, err := NewHTTPClientV3(&Options{
+		Address: "localhost:5000",
+	})
+	assert.NotNil(t, client)
+	assert.NoError(t, err)
+
+	opts := scheme.ScanOptions{}
+	devices, err := client.Scan(opts)
```
Comment (edaniszewski) [inline synse/http_integration_test.go:176]: It may be worth adding some tests to exercise the use of scan options

File: synse/http_integration_test.go [hunk 6]
```diff
@@ -0,0 +1,551 @@
...
+		t.Skip("skipping integration test")
+	}
+
+	client, err := NewHTTPClientV3(&Options{
+		Address: "localhost:5000",
+	})
+	assert.NotNil(t, client)
+	assert.NoError(t, err)
+
+	opts := scheme.TagsOptions{}
```
Comment (edaniszewski) [inline synse/http_integration_test.go:202]: may be worth adding some tests to exercise the tag options

File: synse/http_integration_test.go [hunk 7]
```diff
@@ -0,0 +1,551 @@
...
+	assert.NoError(t, err)
+
+	opts := scheme.ScanOptions{}
+	devices, err := client.Scan(opts)
+	assert.NoError(t, err)
+	assert.Equal(t, 22, len(devices))
+
+	for _, device := range devices {
+		assert.NotEmpty(t, device)
+	}
```
Comment (edaniszewski) [inline synse/http_integration_test.go]: is this block necessary? seems like the same check is being done in the loop below

File: synse/http_integration_test.go [hunk 8]
```diff
@@ -0,0 +1,551 @@
...
+		assert.NotEmpty(t, info.Type)
+		assert.NotEmpty(t, info.Plugin)
+		assert.NotEmpty(t, info.Capabilities.Mode)
+		assert.NotEmpty(t, info.Tags)
+		assert.NotEmpty(t, info.Metadata)
+		assert.Empty(t, info.Capabilities.Write.Actions)
+
+		// TODO - add sort_index to the scheme
+
+		// only led devices have alias
```
Comment (edaniszewski) [inline synse/http_integration_test.go]: see comment above, I'd remove this unless we use a custom emulator device config.

File: synse/http_integration_test.go [hunk 9]
```diff
@@ -0,0 +1,551 @@
...
+		t.Skip("skipping integration test")
+	}
+
+	client, err := NewHTTPClientV3(&Options{
+		Address: "localhost:5000",
+	})
+	assert.NotNil(t, client)
+	assert.NoError(t, err)
+
+	opts := scheme.ReadOptions{}
```
Comment (edaniszewski) [inline synse/http_integration_test.go:262]: would be good to have a test case which exercises the read options

File: synse/http_integration_test.go [hunk 10]
```diff
@@ -0,0 +1,551 @@
...
+	devices, err := client.Read(opts)
```
Comment (edaniszewski) [inline synse/http_integration_test.go]: nitpick, but we aren't getting devices back, so the variable name can be misleading -- could change this to 'readings'

File: synse/http_integration_test.go [hunk 11]
```diff
@@ -0,0 +1,551 @@
...
+	})
+	assert.NotNil(t, client)
+	assert.NoError(t, err)
+
+	opts := scheme.ScanOptions{}
+	devices, err := client.Scan(opts)
+	assert.NoError(t, err)
+	assert.Equal(t, 22, len(devices))
+
+	for _, device := range devices {
```
Comment (edaniszewski) [inline synse/http_integration_test.go]: I'm thinking that it may be worthwhile to define a device configuration for the emulator for these tests so we know the set of devices we are working with that won't change. With that and the fact that we have deterministic IDs, we shouldn't have to iterate over scan results to test ReadDevice (and I think the Info test does the same thing) for every device. Instead we can pick one and verify that the values we get back match the expected. This allows us to use more specific checks than just `NotEmpty` which seems desirable

File: synse/http_integration_test.go [hunk 12]
```diff
@@ -0,0 +1,551 @@
...
+			// "system/type:led",
+			// "system/type:lock",
+			// "system/type:power",
+		},
+	}
+	devices, err := client.Scan(opts)
+	assert.NoError(t, err)
+	assert.Equal(t, 1, len(devices))
+
+	for _, device := range devices {
```
Comment (edaniszewski) [inline synse/http_integration_test.go]: I feel like instead of scanning for devices and iterating over a slice of 1, we can just use a well-known device ID for the write. This gets to the point above about defining a custom device configuration for these tests. We could also define aliases for devices so instead of using some opaque hash string, we could reference devices by a more readable name.

File: synse/http_integration_test.go [hunk 13]
```diff
@@ -0,0 +1,551 @@
...
```
Comment (edaniszewski) [inline synse/http_integration_test.go]: same note as for previous test

File: synse/http_integration_test.go [hunk 14]
```diff
@@ -0,0 +1,551 @@
...
+		assert.NotEmpty(t, device.ID)
+
+		writeData := []scheme.WriteData{
+			{Action: "speed", Data: "101"}, // FIXME - if Data is int, get 500 from server.
+		}
```
Comment (edaniszewski) [inline synse/http_integration_test.go]: this works right now since we are only writing to fan devices. if we include different kinds, this will cause failures for those devices which do not support this action. I think instead of trying to write to all devices, we should focus on writing to one device and verifying all the data from that

**Review by edaniszewski — CHANGES_REQUESTED** [review]
These tests look pretty good, but I think a lot of the assertions are too general. One idea I have for that is to supply a custom device config for the emulator to use, that way we know which devices we are working with and they are not subject to change like the default emulator config would be. In addition to that, scoping some tests to a single device so we can get more specific assertions beyond just NotEmpty/Empty

**Comment by hoanhan101** [comment]
Thanks for the review. Using a custom device config for the emulator seems like a better idea. I will go ahead and update this.

**Comment by hoanhan101** [comment]
I have updated these tests to use only one `led` device. As for the suggestions to exercise scan options, read options and tags options more, I'm not sure if adding more tests for these will help since we only have one device anyway. Other than that, this is ready for a re-review.

File: Makefile [hunk 15]
```diff
@@ -1,7 +1,5 @@
 #
-# synse-client-go
-#
-
+# synse-client-go #
```
Comment (edaniszewski) [inline Makefile]: why the change in the header comment?

File: compose/config/device/fan.yaml [hunk 16]
```diff
@@ -0,0 +1,11 @@
+version: 3
+devices:
+  - type: led
```
Comment (edaniszewski) [inline compose/config/device/fan.yaml]: this is an LED device, but the file is named `fan.yaml`

File: compose/config/tcp/config.yml [hunk 17]
```diff
@@ -0,0 +1,16 @@
+version: 3
```
Comment (edaniszewski) [inline compose/config/tcp/config.yml]: does this file need to be in a `tcp` subdir? it seems like it shouldn't need to be

File: compose/server.yml [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+    - emulator
+
+  emulator:
+    container_name: emulator
+    image: vaporio/emulator-plugin:latest
+    expose:
+    - "5001"
+    command: ["--debug"]
+    volumes:
+      - ./config/tcp:/tmp/config
```
Comment (edaniszewski) [inline compose/server.yml]: the plugin config could just be put in the config dir, so this becomes
```
- ./config/config.yaml:/tmp/config/config.yaml
```

File: synse/http_integration_test.go [hunk 19]
```diff
@@ -0,0 +1,473 @@
...
+
+	client, err := NewHTTPClientV3(&Options{
+		Address: "localhost:5000",
+	})
+	assert.NotNil(t, client)
+	assert.NoError(t, err)
+
+	version, err := client.Version()
+	assert.NoError(t, err)
+	assert.Equal(t, "3.0.0-alpha.3", version.Version)
```
Comment (edaniszewski) [inline synse/http_integration_test.go]: this is a case where just asserting that it is not empty is fine, since this value will change based on whatever the `:latest` tag is. the alternative is to pin the tag to a known version in the compose file, but I think that could lead to problems down the line

File: synse/http_integration_test.go [hunk 20]
```diff
@@ -0,0 +1,473 @@
...
+	assert.Equal(t, "4032ffbe-80db-5aa5-b794-f35c88dff85c", plugin.ID)
+	assert.True(t, plugin.Active)
+
+	assert.Equal(t, "emulator:5001", plugin.Network.Address)
+	assert.Equal(t, "tcp", plugin.Network.Protocol)
+	assert.Equal(t, "3.0.0-alpha.3", plugin.Version.PluginVersion)
+	assert.Equal(t, "3.0.0-alpha.1", plugin.Version.SDKVersion)
+	assert.NotEmpty(t, plugin.Version.BuildDate)
+	assert.Equal(t, "4234777", plugin.Version.GitCommit)
+	assert.Equal(t, "3.0.0-alpha.3", plugin.Version.GitTag)
```
Comment (edaniszewski) [inline synse/http_integration_test.go]: the git tag / commit can just be an empty check since this will change depending on what the `:latest` image is

File: synse/http_integration_test.go [hunk 21]
```diff
@@ -0,0 +1,473 @@
...
+	}
+	writes, err := client.WriteAsync("f041883c-cf87-55d7-a978-3d3103836412", writeData)
+	assert.NoError(t, err)
+	assert.Equal(t, 2, len(writes))
+
+	stateWrite := writes[0]
+	assert.NotEmpty(t, stateWrite.ID)
+	assert.Equal(t, "f041883c-cf87-55d7-a978-3d3103836412", stateWrite.Device)
+	assert.Equal(t, "state", stateWrite.Context.Action)
+	// assert.Equal(t, "on", stateWrite.Context.Data) // FIXME - reflected data isn't decoded yet
```
Comment (edaniszewski) [inline synse/http_integration_test.go]: with the latest alpha release of synse server, this should have been fixed

**Review by edaniszewski — COMMENTED** [review]
thanks for the changes! i think it looks a lot more extensive now. 

one question I have is why we only have one device in the custom emulator configuration? It could be done in a follow-up PR, but it seems like it would be good to add in other devices, which would also enable some basic integration testing of various options. I think what I'm most interested in are the options for filtering based on tags -- for that, i think we'd only need maybe 3 temperature devices defined.

File: Makefile (hunk 15 above)
Comment (hoanhan101) [reply Makefile]: it was a mistake! thanks for catching this.

File: compose/config/device/fan.yaml (hunk 16 above)
Comment (hoanhan101) [reply compose/config/device/fan.yaml]: this too

**Comment by hoanhan101** [comment]
Updated some changes regarding to the reviews. Will add in some other devices in a follow-up PR. Is that okay with you, @edaniszewski?

**Comment by edaniszewski** [comment]
thanks for the updates! sorry I didn't re-review sooner, I must have missed the updates. the changes look good, but CI is failing on test with the following error:

```
+ CGO_ENABLED=0 go test -covermode=atomic ./...
go: disabling cache (/.cache/go-build) due to initialization failure: mkdir /.cache: permission denied
go: cannot use modules with build cache disabled
script returned exit code 1
```

@lazypower have you seen this before/have any ideas what it might be? I don't think there have been any changes to the CI setup here for a while and tests were passing before so I think/hope its just some intermittent CI failure?

**Comment by lazypower** [comment]
TBH it looks like a resurfacing of https://github.com/golang/go/issues/26809

per this issue:

> The GOCACHE env variable will be phased out out in 1.12, and cannot be used with go modules > > enabled in 1.11. https://tip.golang.org/doc/go1.11#gocache
> I presume you want to disable the test caching, in which case you can use `go test -count=1 ...`

**Comment by lazypower** [comment]
Ah one second, let me investigate the pipeline more than I did on that last comment (which was deleted, if you missed it). I didn't notice the .cache mkdir perm failure. This is likely, again, user permissions when using docker on host. 

Another reason why we need to be pushing towards clean room testing envs in kubernetes vs using this shared workspace model where UID's are fast changing and causing problems in workspace cache because care isn't used to clean them up, mostly because nobody expects UID/GID changes on files when going from in/out of the container.

but thats for another blog post
